File: findMedianSortedArrays.kt
fun select(l1: IntArray, l2: IntArray, k: Int): Int {
    if (l1.isEmpty()) {
        return l2[k]
    }
    if (l2.isEmpty()) {
        return l1[k]
    }
    val pivot = l1[l1.size / 2]
    val index = l2.binarySearch(pivot)
    val ltPivotInL1Cnt = l1.size / 2
    val ltPivotInL2Cnt = if (index >= 0) index else -index - 1
    val ltPivotCnt = ltPivotInL1Cnt + ltPivotInL2Cnt
    return when {
        k < ltPivotCnt -> {
            val subL1 = l1.sliceArray(0 until ltPivotInL1Cnt)
            val subL2 = l2.sliceArray(0 until ltPivotInL2Cnt)
            select(subL1, subL2, k)
        }
        k > ltPivotCnt -> {
            val subL1 = l1.sliceArray(ltPivotInL1Cnt + 1 until l1.size)
            val subL2 = l2.sliceArray(ltPivotInL2Cnt until l2.size)
            select(subL1, subL2, k - ltPivotCnt - 1)
        }
        else -> pivot
    }
}

fun findMedianSortedArrays(nums1: IntArray, nums2: IntArray): Double {
    val len = nums1.size + nums2.size
    return when (len % 2) {
        0 -> {
            val ltMed = select(nums1, nums2, len / 2 - 1)
            val rtMed = select(nums1, nums2, len / 2)
            (ltMed + rtMed).toDouble() / 2
        }
        else -> select(nums1, nums2, len / 2).toDouble()
    }
}

File: findMedianSortedArrays.c
int binarySearch(int* nums, int size, int target) {
    int l = 0;
    int r = size;
    while (l < r) {
        int m = (l + r) / 2;
        if (nums[m] == target) {
            return m;
        } else if (nums[m] < target) {
            l = m + 1;
        } else {
            r = m;
        }
    }
    return l;
}

int select(int* nums1, int nums1Size, int* nums2, int nums2Size, int k) {
    if (nums1Size == 0) {
        return nums2[k];
    }
    if (nums2Size == 0) {
        return nums1[k];
    }
    int pivot = nums1[nums1Size / 2];
    int nums1Cnt = nums1Size / 2;
    int nums2Cnt = binarySearch(nums2, nums2Size, pivot);
    int cnt = nums1Cnt + nums2Cnt;
    if (cnt < k) {
        return select(
            &nums1[nums1Cnt + 1],
            nums1Size - nums1Cnt - 1,
            &nums2[nums2Cnt],
            nums2Size - nums2Cnt,
            k - cnt - 1
        );
    } else if (cnt > k) {
        return select(nums1, nums1Cnt, nums2, nums2Cnt, k);
    } else {
        return pivot;
    }
}

double findMedianSortedArrays(int* nums1, int nums1Size, int* nums2, int nums2Size) {
    int size = nums1Size + nums2Size;
    if (size % 2 == 0) {
        int l = select(nums1, nums1Size, nums2, nums2Size, size / 2 - 1);
        int r = select(nums1, nums1Size, nums2, nums2Size, size / 2);
        return ((double) (l + r)) / 2;
    } else {
        return select(nums1, nums1Size, nums2, nums2Size, size / 2);
    }
}
